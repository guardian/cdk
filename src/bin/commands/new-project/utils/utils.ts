import camelCase from "lodash.camelcase";
import upperFirst from "lodash.upperfirst";

export const pascalCase = (str: string): string => {
  return upperFirst(camelCase(str));
};

export interface Name {
  kebab: string;
  pascal: string;
}
