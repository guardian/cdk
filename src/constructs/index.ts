export * as acm from "./acm";
export * as autoscaling from "./autoscaling";
export * as cloudwatch from "./cloudwatch";
export * as core from "./core";
export * as dns from "./dns";
export * as dynamodb from "./dynamodb";
export * as ec2 from "./ec2";
export * as ecs from "./ecs";
export * as iam from "./iam";
export * as kinesis from "./kinesis";
export * as lambda from "./lambda";
export * as loadbalancing from "./loadbalancing";
export * as rds from "./rds";
export * as s3 from "./s3";
export * as ses from "./ses";
export * as vpc from "./vpc";
export * as root from "./root";
