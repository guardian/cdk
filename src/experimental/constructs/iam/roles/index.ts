export * from "./fastly-kinesis-log";
export * from "./cross-account-role";
