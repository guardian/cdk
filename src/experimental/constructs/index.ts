export * from "./error-budget-alarm";
export * from "./root";
