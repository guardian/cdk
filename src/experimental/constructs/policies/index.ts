export * from "./kinesis-put-records";
