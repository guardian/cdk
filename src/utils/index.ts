export * as array from "./array";
export * as exec from "./exec";
export * as git from "./git";
export * as math from "./math";
export * as singleton from "./singleton";
