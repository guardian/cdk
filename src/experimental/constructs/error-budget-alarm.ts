import { Duration } from "aws-cdk-lib";
import type { CfnCompositeAlarm, IMetric } from "aws-cdk-lib/aws-cloudwatch";
import {
  Alarm,
  AlarmRule,
  AlarmState,
  ComparisonOperator,
  CompositeAlarm,
  MathExpression,
  TreatMissingData,
} from "aws-cdk-lib/aws-cloudwatch";
import { SnsAction } from "aws-cdk-lib/aws-cloudwatch-actions";
import type { ITopic } from "aws-cdk-lib/aws-sns";
import { Topic } from "aws-cdk-lib/aws-sns";
import { Construct } from "constructs";
import type { GuStack } from "../../constructs/core";

export interface ErrorBudgetAlarmProps {
  sloName: string;
  sloTarget: number;
  badEvents: IMetric;
  validEvents: IMetric;
  snsTopicNameForAlerts: string;
}

interface BurnRate {
  speed: "Fast" | "Medium" | "Slow";
  burnRate: number;
}

interface BurnRateMonitoring extends BurnRate {
  longPeriod: Duration;
  shortPeriod: Duration;
}

interface MonitorBurnRateForPeriodProps {
  alarmName: string;
  burnRateSpeed: BurnRate;
  period: Duration;
  errorBudget: number;
  badEvents: IMetric;
  validEvents: IMetric;
}

interface CompositeBurnRateAlarmProps {
  sloName: string;
  burnRateMonitoring: BurnRateMonitoring;
  errorBudget: number;
  badEvents: IMetric;
  validEvents: IMetric;
  snsTopic: ITopic;
  suppressorAlarm?: CompositeBurnRateAlarm;
}

/**
 * Use this construct to create alarms that will notify you when you are at risk of missing your SLO targets.
 *
 * These alarms follow the recommendations in the [SRE Workbook](https://sre.google/workbook/alerting-on-slos/) to
 * implement Multiwindow, Multi-Burn-Rate Alerts.
 *
 * The CloudWatch implementation of this strategy creates nine separate CloudWatch alarms. The main alarms that users
 * need to care about are the three composite alarms. These track different burn rates of error budget consumption:
 * fast (e.g. complete outages), medium (e.g. moderate error percentage over hours, rather than minutes) and slow (e.g.
 * a small percentage of errors sustained over a much longer period). Multiple burn rates are used to ensure that
 * we are able to pick up all scenarios which could threaten the SLO, without interrupting developers with an
 * unnecessary level of urgency.
 *
 * Each composite alarm creates two child alarms; these are used to support multiple windows (or periods in CloudWatch
 * terminology). A composite alarm is configured to send notifications if both of its child alarms are in an alarm state.
 * The child alarms themselves have no actions/notifications configured and users should not need to interact with them
 * directly. Multiple windows are helpful for accuracy and reset time. The long window helps us to avoid sending an
 * alert in scenarios where no intervention is required, for example a very brief spike in errors that is resolved
 * automatically (e.g. when AWS replaces a 'bad' EC2 instance). The short window helps to ensure that the alert is reset
 * (or moved back into OK state, in CloudWatch terminology) promptly once a problem has been resolved.
 *
 * The composite alarms have an awareness of priority, meaning that low priority alarm notifications are suppressed if
 * a higher priority alarm is already firing. That is, if you receive an alert about fast error budget consumption, you
 * should not receive an alert from the medium or slow versions of the alarm.
 */
export class GuErrorBudgetAlarmExperimental extends Construct {
  constructor(scope: GuStack, props: ErrorBudgetAlarmProps) {
    super(scope, props.sloName); // The assumption here is that `sloName` will be unique (per stack)

    const errorBudget = 1 - props.sloTarget;

    const snsTopic: ITopic = Topic.fromTopicArn(
      scope,
      `SnsSloAlarmsFor${props.sloName}`,
      `arn:aws:sns:${scope.region}:${scope.account}:${props.snsTopicNameForAlerts}`
    );

    const fastBurnRate: BurnRateMonitoring = {
      speed: "Fast",
      burnRate: 14.4,
      longPeriod: Duration.hours(1),
      shortPeriod: Duration.minutes(5),
    };

    const fastBurnAlarm = new CompositeBurnRateAlarm(scope, {
      sloName: props.sloName,
      badEvents: props.badEvents,
      burnRateMonitoring: fastBurnRate,
      errorBudget,
      snsTopic,
      validEvents: props.validEvents,
    });

    const mediumBurnRateMonitoring: BurnRateMonitoring = {
      speed: "Medium",
      burnRate: 6,
      longPeriod: Duration.hours(6),
      shortPeriod: Duration.minutes(30),
    };

    const mediumBurnRateAlarm = new CompositeBurnRateAlarm(scope, {
      sloName: props.sloName,
      badEvents: props.badEvents,
      burnRateMonitoring: mediumBurnRateMonitoring,
      errorBudget,
      snsTopic,
      validEvents: props.validEvents,
      suppressorAlarm: fastBurnAlarm,
    });

    const slowBurnRate: BurnRateMonitoring = {
      speed: "Slow",
      burnRate: 3,
      longPeriod: Duration.days(1),
      shortPeriod: Duration.hours(2),
    };

    new CompositeBurnRateAlarm(scope, {
      sloName: props.sloName,
      badEvents: props.badEvents,
      burnRateMonitoring: slowBurnRate,
      errorBudget,
      // TODO - use a different topic for tickets or figure out how to route notifications vs tickets using a single topic
      snsTopic,
      validEvents: props.validEvents,
      suppressorAlarm: mediumBurnRateAlarm,
    });
  }
}

class MonitorBurnRateForPeriod extends Alarm {
  constructor(scope: GuStack, props: MonitorBurnRateForPeriodProps) {
    const undesirableErrorBudgetConsumption = props.errorBudget * props.burnRateSpeed.burnRate;
    super(scope, props.alarmName, {
      alarmName: props.alarmName,
      metric: new MathExpression({
        expression: "badEvents/validEvents",
        label: "Observed failure rate",
        usingMetrics: {
          badEvents: props.badEvents,
          validEvents: props.validEvents,
        },
        period: props.period,
      }),
      evaluationPeriods: 1,
      threshold: undesirableErrorBudgetConsumption,
      comparisonOperator: ComparisonOperator.GREATER_THAN_THRESHOLD,
      treatMissingData: TreatMissingData.NOT_BREACHING,
    });
  }
}

class CompositeBurnRateAlarm extends CompositeAlarm {
  constructor(scope: GuStack, props: CompositeBurnRateAlarmProps) {
    const alarmName = `${props.sloName}${props.burnRateMonitoring.speed}BurnCompositeAlarm`;
    const sharedIndividualAlarmProps = {
      badEvents: props.badEvents,
      burnRateSpeed: {
        speed: props.burnRateMonitoring.speed,
        burnRate: props.burnRateMonitoring.burnRate,
      },
      errorBudget: props.errorBudget,
      validEvents: props.validEvents,
    };
    super(scope, alarmName, {
      alarmRule: AlarmRule.allOf(
        AlarmRule.fromAlarm(
          new MonitorBurnRateForPeriod(scope, {
            ...sharedIndividualAlarmProps,
            alarmName: `ChildAlarmLongPeriod${alarmName}`,
            period: props.burnRateMonitoring.longPeriod,
          }),
          AlarmState.ALARM
        ),
        AlarmRule.fromAlarm(
          new MonitorBurnRateForPeriod(scope, {
            ...sharedIndividualAlarmProps,
            alarmName: `ChildAlarmShortPeriod${alarmName}`,
            period: props.burnRateMonitoring.shortPeriod,
          }),
          AlarmState.ALARM
        )
      ),
    });
    this.addAlarmAction(new SnsAction(props.snsTopic));
    if (props.suppressorAlarm) {
      const cfnAlarm = this.node.defaultChild as CfnCompositeAlarm;
      cfnAlarm.actionsSuppressor = props.suppressorAlarm.alarmArn;
      cfnAlarm.actionsSuppressorWaitPeriod = 120;
      cfnAlarm.actionsSuppressorExtensionPeriod = props.burnRateMonitoring.shortPeriod.toSeconds();
    }
  }
}
