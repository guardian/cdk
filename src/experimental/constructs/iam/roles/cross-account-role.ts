import { ArnPrincipal } from "aws-cdk-lib/aws-iam";
import type { GuStack } from "../../../../constructs/core";
import type { GuRoleProps } from "../../../../constructs/iam";
import { GuRole } from "../../../../constructs/iam";

export interface GuCrossAccountRoleExperimentalProps extends Omit<GuRoleProps, "assumedBy"> {
  /**
   * The name of the role that will assume this new GuCrossAccountRoleExperimental.
   * It should be the full name including the stage (if appropriate).
   */
  nameOfRoleWhichCanAssumeThisRole: string;

  /**
   * The AWS account ID associated with the resource that will consume this new GuCrossAccountRoleExperimental.
   *
   * If this construct is being used on a public repo it would be advisable to use
   * [private-infrastructure-config](https://github.com/guardian/private-infrastructure-config) so that any account ID
   * is not publicly available.
   */
  accountId: string;
}

/**
 * A construct to create a cross account role.
 *
 * In order to use this construct the name of the role that will assume this cross account role must be provided, along
 * with the corresponding AWS account ID of the assuming role.
 *
 * The resulting role can only be assumed by the specified role in the given AWS account. When this role has been
 * created it can be extended to grant permissions for specific actions, allowing a resource in one AWS account to
 * perform actions in another.
 */
export class GuCrossAccountRoleExperimental extends GuRole {
  constructor(scope: GuStack, id: string, props: GuCrossAccountRoleExperimentalProps) {
    const { nameOfRoleWhichCanAssumeThisRole, accountId } = props;

    super(scope, id, {
      assumedBy: new ArnPrincipal(`arn:aws:iam::${accountId}:role/${nameOfRoleWhichCanAssumeThisRole}`),
      ...props,
    });
  }
}
