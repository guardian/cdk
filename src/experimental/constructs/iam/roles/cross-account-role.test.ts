import { Template } from "aws-cdk-lib/assertions";
import { ServicePrincipal } from "aws-cdk-lib/aws-iam";
import { GuRole } from "../../../../constructs/iam";
import { simpleGuStackForTesting } from "../../../../utils/test";
import { GuCrossAccountRoleExperimental } from "./cross-account-role";

describe("The GuCrossAccountRoleExperimental construct", () => {
  it("can create a cross account role", () => {
    const stack = simpleGuStackForTesting();
    new GuCrossAccountRoleExperimental(stack, "testCrossAccountRole", {
      nameOfRoleWhichCanAssumeThisRole: "nameOfRoleInOtherAccountWhichCanAssumeThis-STAGE",
      roleName: "crossAccountRole",
      accountId: "1234"
    });
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });

  it("can create a cross account role that can be assumed by a service in another account", () => {
    const stackThatCreatesTheRole = simpleGuStackForTesting();
    new GuCrossAccountRoleExperimental(stackThatCreatesTheRole, "testCrossAccountRole", {
      nameOfRoleWhichCanAssumeThisRole: "nameOfRoleInOtherAccountWhichCanAssumeThisNewlyCreatedOne-CODE",
      roleName: "crossAccountRole",
      accountId: "1234"
    });

    Template.fromStack(stackThatCreatesTheRole).hasResourceProperties("AWS::IAM::Role", {
      RoleName: "crossAccountRole",
      AssumeRolePolicyDocument: {
        Statement: [{
          Action: "sts:AssumeRole",
          Effect: "Allow",
          Principal: {
            AWS: "arn:aws:iam::1234:role/nameOfRoleInOtherAccountWhichCanAssumeThisNewlyCreatedOne-CODE"
          }
        }]
      }
    })

    const stackThatAssumesTheRole = simpleGuStackForTesting();
    new GuRole(stackThatAssumesTheRole, "idForRole", {
      assumedBy: new ServicePrincipal("ec2.amazonaws.com"),
      roleName: "nameOfRoleInOtherAccountWhichCanAssumeThisNewlyCreatedOne-CODE"
    });

    Template.fromStack(stackThatAssumesTheRole).hasResourceProperties("AWS::IAM::Role", {
      RoleName: "nameOfRoleInOtherAccountWhichCanAssumeThisNewlyCreatedOne-CODE",
      AssumeRolePolicyDocument: {
        Statement: [{
          Action: "sts:AssumeRole",
          Effect: "Allow",
          Principal: {
            Service: "ec2.amazonaws.com"
          }
        }]
      }
    });
  })
});

