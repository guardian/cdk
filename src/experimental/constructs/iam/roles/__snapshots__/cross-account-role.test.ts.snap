// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuCrossAccountRoleExperimental construct can create a cross account role 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuStack",
      "GuCrossAccountRoleExperimental",
    ],
    "gu:cdk:version": "TEST",
  },
  "Resources": {
    "testCrossAccountRole975BA7C0": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": "arn:aws:iam::1234:role/nameOfRoleInOtherAccountWhichCanAssumeThis-STAGE",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": "crossAccountRole",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
